### PR TITLE
Placeholders in service-url for dynamic parameters (#8)

### PR DESCRIPTION
**Pull request by dprzybyl**
---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/RequestPlaceholder.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.stereotype.Component;
+
+import io.vertx.core.http.HttpServerRequest;
+
+@Component
+public class RequestPlaceholder implements Placeholder {
+
+    @Override
+    public String getValue(final HttpServerRequest request, final String placeholder) {
+        return Arrays.stream(Strategy.values()).filter(x -> StringUtils.startsWith(placeholder, x.prefix))
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/RequestPlaceholder.java]: Please rename `x` to `strategy` or something more meaningful.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/Placeholder.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+package com.cognifide.knotx.placeholder;
+
+import io.vertx.core.http.HttpServerRequest;
+
+public interface Placeholder {
+
+	String getValue(HttpServerRequest request, String placeholder);
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/Placeholder.java]: Please add javadoc to interface method.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriPlaceholder.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+public class UriPlaceholder implements Placeholder {
+
+    private UriInfo uriInfo;
+
+    @Override
+    public String getValue(HttpServerRequest request, String placeholder) {
+        if (uriInfo == null) {
+            uriInfo = new UriInfo(request.uri());
+        }
+        return Arrays.stream(Strategy.values()).filter(x -> StringUtils.startsWith(placeholder, x.prefix))
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriPlaceholder.java]: Please rename `x` to `strategy` or something more meaningful.

File: knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            "/a/b.s1.html/c/d#f,            /a/b, s1,    html, /c/d",
+            "/a/b.s1.s2.html/c/d#f,         /a/b, s1.s2, html, /c/d",
+            "/a/b.html/c/d.s.txt#f,         /a/b, null,  html, /c/d.s.txt",
+            "/a/b.s1.html/c/d.s.txt#f,      /a/b, s1,    html, /c/d.s.txt",
+            "/a/b.s1.s2.html/c/d.s.txt#f,   /a/b, s1.s2, html, /c/d.s.txt"
+    })
+    public void testExtractingValuesFromUri(String uri, String path, String selectorString, String extension,
+            String suffix) {
+        final UriInfo uriInfo = new UriInfo(uri);
+        boolean result = StringUtils.equals(uriInfo.getPath(), path);
```
Comment (Skejven) [inline knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java]: Can you provide separate assertion for each check?

**Comment by Skejven** [comment]
Please update documentation and examples in `README.md`.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+package com.cognifide.knotx.placeholder;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class UriInfo {
+
+    private static final Pattern URI_PATTERN = Pattern
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java]: Please write short javadoc about this pattern.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            suffix = matcher.group(6);
+        }
+    }
+
+    public String getPath() {
+        return path;
+    }
+
+    public String getPathPart(int index) {
+        return pathparts != null && 0 <= index && index < pathparts.length ? pathparts[index] : null;
```
Comment (toniedzwiedz) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java]: Looks complex enough to use parentheses

`return (pathparts != null && 0 <= index && index < pathparts.length) ? pathparts[index] : null;`

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriPlaceholderResolver.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    public UriPlaceholderResolver(HttpServerRequest request) {
+        this.request = request;
+    }
+
+    @Override
+    public String resolvePlaceholder(String placeholder) {
+        if (!cache.containsKey(placeholder)) {
+            final String value = placeholders.stream().map(bean -> bean.getValue(request, placeholder))
+                    .filter(Objects::nonNull).findFirst().orElse(null);
+            cache.put(placeholder, value);
```
Comment (toniedzwiedz) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriPlaceholderResolver.java]: Do we really want to put `null` values in the map?

File: knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            "/a/b.s1.html/c/d#f,            /a/b, s1,    html, /c/d",
+            "/a/b.s1.s2.html/c/d#f,         /a/b, s1.s2, html, /c/d",
+            "/a/b.html/c/d.s.txt#f,         /a/b, null,  html, /c/d.s.txt",
+            "/a/b.s1.html/c/d.s.txt#f,      /a/b, s1,    html, /c/d.s.txt",
+            "/a/b.s1.s2.html/c/d.s.txt#f,   /a/b, s1.s2, html, /c/d.s.txt"
+    })
+    public void testExtractingValuesFromUri(String uri, String path, String selectorString, String extension,
+            String suffix) {
+        final UriInfo uriInfo = new UriInfo(uri);
+        Assert.assertTrue(StringUtils.equals(uriInfo.getPath(), path));
```
Comment (toniedzwiedz) [inline knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java]: This isn't very readable. How about

    assertThat(uriInfo.getPath(), is(path));

Not sure if we have Hamcrest or AssertJ as a dependency already but if not, I think it's worth introducing.

**Comment by marcinczeczko** [comment]
PR need to be modified to new codebase merged to master

**Comment by Skejven** [comment]
Closing as outdated. Continuation of this task in #8

**Pull request closed without merging**